### PR TITLE
Script to download training data, images and convert data to csv files

### PR DESCRIPTION
**Pull request by samiyaakhtar**
- Deleted existing (unused) api /taggedimages and split it into api/images and api/tags.  
- Downloading images to a folder and converting the metadata into csv recognizable by scripts
- Optimized some existing queries to fetch training data for tagging complete images by getting all data in one query instead of making a separate query for each image id.

Potentially might need to tweak the formats in csv/foldernames if needed

### CODE REVIEW

File: functions/pipeline/shared/db_access/db_access_v2.py [hunk 1]
```diff
@@ -130,19 +130,16 @@ def add_new_images(self,list_of_image_infos, user_id):
...
+    def get_tag_complete_images(self, user_id):
         tag_complete_images = {}
         try:
             conn = self._db_provider.get_connection()
             try:
                 cursor = conn.cursor()
                 query = ("SELECT b.ImageId, b.ImageLocation, a.TagStateId FROM Image_Tagging_State a "
-                        "JOIN Image_Info b ON a.ImageId = b.ImageId WHERE a.TagStateId = {1} order by "
-                        "a.createddtim DESC limit {0}")
-                cursor.execute(query.format(number_of_images, ImageTagState.COMPLETED_TAG))
```
Comment (mtarng) [inline functions/pipeline/shared/db_access/db_access_v2.py:145]: Not limiting could cause long runtimes over the limit (300 secs?) or cause the final return payloads to grow over what is allowed through our API gateway.

Comment (samiyaakhtar) [reply functions/pipeline/shared/db_access/db_access_v2.py:145]: Right, we decided to use all images for training (for now) so that's why I removed the limit, @andrebriggs should I bring a limit back?

File: functions/pipeline/taggedimages/__init__.py [hunk 2]
```diff
@@ -35,19 +35,24 @@ def main(req: func.HttpRequest) -> func.HttpResponse:
...
+            obj = list(image_id_to_urls.keys())
+            image_id_to_image_tags = data_access.get_image_tags_for_image_ids(obj)
 
             existing_classifications_list = data_access.get_existing_classifications()
 
             vott_json = create_starting_vott_json(image_id_to_urls, image_id_to_image_tags, existing_classifications_list)
 
-            return_body_json = {"imageUrls": image_urls,
-                                "vottJson": vott_json}
+            return_body_json = {"imageUrls": image_id_to_urls,
```
Comment (andrebriggs) [inline functions/pipeline/taggedimages/__init__.py]: @samiyaakhtar It seems like we are overloading this function. The "taggedImage" API is currently only used to create a VOTT JSON file right? I see it is now being called in the Training.py I think we should have a separate API or some sort of flag on the API

File: train/training.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.y1 = y1
+        self.y2 = y2
+        self.height = height
+        self.width = width
+        self.name = name
+
+
+def train(config, num_images):
+    # First, downloxad vott json for tagging complete images
+    vott_json = download_vott_json(config, num_images)
```
Comment (andrebriggs) [inline train/training.py]: @samiyaakhtar for training VOTT json isn't used at all. Perhaps we can change the language here. I think this is a symptom of overloading the TaggedImage API with extra data

File: train/training.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    file_location = Config.initialize_tagging_location(config)
+
+    # Grab tagged and totag images from the blob storage
+    download_images(vott_json, str(file_location))
+
+    # create csv file from this data
+    convert_to_csv(vott_json, str(file_location))
+
+
+def download_images(vott_json, file_location):
```
Comment (andrebriggs) [inline train/training.py]: @samiyaakhtar I need to check the existing code but I'm not sure if images are downloaded into separate folders based on tag status.

File: functions/pipeline/shared/db_access/db_access_v2.py [hunk 5]
```diff
@@ -155,6 +152,98 @@ def get_tag_complete_images(self, number_of_images, user_id):
...
+                    ids += str(id) + ','
+                ids = ids[:-1]
+                query = ("select imageid, originalimagename, imagelocation, height, width, createdbyuser from image_info where imageid IN ({0});")
+                cursor.execute(query.format(ids))
+                logging.debug("Got image info back for image_id={}".format(image_ids))
+
+                images_info = []
+                for row in cursor:
+                    info = {}
+                    info['height'] = row[3]
```
Comment (andrebriggs) [inline functions/pipeline/shared/db_access/db_access_v2.py:174]: @samiyaakhtar I know it's faster create a list of dictionaries but I think creating a Class that defined the properties would allow you to share code on the backend (Data access and function) with the frontend (CLI client). This would create a contract. You can deserialize the JSON to a Class and serialize a Class to JSON. See an example here: https://stackoverflow.com/a/15477213

Something to think about for the future

File: functions/pipeline/shared/db_access/db_access_v2.py [hunk 6]
```diff
@@ -155,6 +152,98 @@ def get_tag_complete_images(self, number_of_images, user_id):
             conn.close()
         return tag_complete_images
 
+    def get_ready_to_tag_images(self, number_of_images, user_id):
```
Comment (andrebriggs) [inline functions/pipeline/shared/db_access/db_access_v2.py]: @samiyaakhtar This is duplicated code with get_tag_complete_images. The only change is the status type. Perhaps we can cut down on the duplication?

File: functions/pipeline/shared/db_access/db_access_v2.py (hunk 1 above)
Comment (andrebriggs) [reply functions/pipeline/shared/db_access/db_access_v2.py:145]: @samiyaakhtar sure for now let's set a limit for now. Maybe choose 200 or something

**Review by andrebriggs — CHANGES_REQUESTED** [review]
@samiyaakhtar I think we should really have an API named "Label". Should take a parameter that specifies the tags. Overloading the Get Tagged Images API with extra data you're not using is a signal that maybe it should be split up. We also need to grab the box and image confidence columns [here](https://github.com/CatalystCode/active-learning-detect/blob/1c662a77f2ceb79840b36c56a50a645f8125115c/db/tables/110_prediction_info.sql#L5). Let's chat tomorrow about this

File: functions/pipeline/taggedimages/__init__.py (hunk 2 above)
Comment (samiyaakhtar) [reply functions/pipeline/taggedimages/__init__.py]: I am overloading this one for now because it wasn't being used anywhere else, and I plan to separate it out but wanted to get something out there for us to have for testing the training pipeline. I can separate it out as part of this PR itself.

File: train/training.py (hunk 4 above)
Comment (samiyaakhtar) [reply train/training.py]: Currently they end up in tagged/ and totag/ let me know what this should change to

File: functions/pipeline/shared/db_access/db_access_v2.py (hunk 6 above)
Comment (samiyaakhtar) [reply functions/pipeline/shared/db_access/db_access_v2.py]: There's difference in the limit n, which is why I created a separate function for now, but I will merge them.

File: functions/pipeline/images/__init__.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            body=json.dumps({"error": "tag status not specified"})
+        )
+    else:
+        try:
+            # DB configuration
+            data_access = ImageTagDataAccess(get_postgres_provider())
+            user_id = data_access.create_user(user_name)
+
+            # Get images info
+            ready_to_tag_images = data_access.get_images_by_tag_status(user_id, tag_status, image_count)
```
Comment (andrebriggs) [inline functions/pipeline/images/__init__.py:47]: @samiyaakhtar these aren't "ready_to_tag_images" they can be anything, right? Perhaps change the name from ready_to_tag_images to requested_images?

File: functions/pipeline/images/__init__.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            headers=headers,
+            body=json.dumps({"error": "invalid userName given or omitted"})
+        )
+    elif not image_count:
+        return func.HttpResponse(
+            status_code=400,
+            headers=headers,
+            body=json.dumps({"error": "image count not specified"})
+        )
+    elif not tag_status:
```
Comment (andrebriggs) [inline functions/pipeline/images/__init__.py:34]: @samiyaakhtar Do you think we should support multiple tag statuses in the param?

File: functions/pipeline/images/__init__.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    else:
+        try:
+            # DB configuration
+            data_access = ImageTagDataAccess(get_postgres_provider())
+            user_id = data_access.create_user(user_name)
+
+            # Get images info
+            ready_to_tag_images = data_access.get_images_by_tag_status(user_id, tag_status, image_count)
+
+            image_infos = data_access.get_image_info_for_image_ids(list(ready_to_tag_images.keys()))
```
Comment (andrebriggs) [inline functions/pipeline/images/__init__.py:49]: @samiyaakhtar one optimization might be to create a method on the data access layer that combines data_access.get_images_by_tag_status and data_access.get_image_info_for_image_ids so you don't have to make two calls from the. We should be able to create a query that gives us all the columns we want in one shot.

File: functions/pipeline/images/__init__.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    image_count = int(req.params.get('imageCount'))
+    user_name = req.params.get('userName')
+    tag_status = req.params.get('tagStatus')
+
+    # setup response object
+    headers = {
+        "content-type": "application/json"
+    }
+    if not user_name:
```
Comment (andrebriggs) [inline functions/pipeline/images/__init__.py:22]: @samiyaakhtar one scenario might be that the CLI calls our Labels API and gets a list of tags that include the image Id. Next the CLI calls this Images API and does a GET call with the list of image Ids as a param. This Images API returns a list of Image Info metadata (url, height width, etc). What do you think about supporting this scenario?

File: functions/pipeline/shared/db_access/db_access_v2.py [hunk 11]
```diff
@@ -39,10 +38,11 @@ def __init__(self, image_id, x_min, x_max, y_min, y_max, classification_names):
 
 # Vott tags have image height & width data as well.
 class VottImageTag(ImageTag):
-    def __init__(self, image_id, x_min, x_max, y_min, y_max, classification_names, image_height, image_width):
+    def __init__(self, image_id, x_min, x_max, y_min, y_max, classification_names, image_height, image_width, image_location):
```
Comment (andrebriggs) [inline functions/pipeline/shared/db_access/db_access_v2.py:41]: @samiyaakhtar I don't think we should overload this. The training scripts have a different need. They need the box and image confidence scores. This class's usage seems to be for generating VOTT json that is needed on download. I suggest renaming this class or creating another class with all these fields plus box and image confidence. Really the word "VOTT" shouldn't exist anywhere in our db data access code cc @mtarng

File: train/training.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    print("Downloaded images into " + file_location + "/AllImages/")
+
+
+def download_data_for_training(config, num_images):
+    print("Downloading data for training, this may take a few moments...")
+    # Download n images that are ready to tag
+    query = {
+        "userName": config.get('tagging_user'),
+        "imageCount": num_images,
+        "tagStatus": 1
```
Comment (andrebriggs) [inline train/training.py:53]: @samiyaakhtar Use the enum

File: train/training.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    url = config.get('url') + '/api/images'
+    response = requests.get(url, params=query)
+    to_tag_image_info = response.json()
+    image_urls_to_download = [info['location'] for info in to_tag_image_info]
+
+    # Download upto 200 images that have been tagged, for training
+    query['tagStatus'] = 3
+    query['imageCount'] = 200
+    url = config.get('url') + '/api/tags'
```
Comment (andrebriggs) [inline train/training.py:63]: @samiyaakhtar I mentioned this in another comment but it might make more sense to be driven by the labels (what you're calling tags) then download the images based upon the image ids in the tags. The current way doesn't seem to guarantee that the image and label data will be correlated.

**Review by andrebriggs — COMMENTED** [review]
@samiyaakhtar We should probably just check this in to bring it to a conclusion and then we can split up some of the work. What I'm thinking in terms of tasks:

+ Create a method on the data access layer that returns a list of objects that have the following
 columns: ImageName, ClassificationName, XMin, XMax, YMin, YMax, Height, Width, Box Confidence, Image confidence
+ Update the Images API to have the following parameters: ImageIds (optional), TagStatus(optional but ignored if ImageIds param is present), ImageCount (optional but required if TagStatus param is present and only used if TagStatus param is present) 
+ Update Tags (I think Labels is more accurate but up to you) API to accept one to many tag statuses as a param.
+ Update Tags API to use new data access method that combines the two data access methods it is currently using.
+ Update Tags API to stop using create_starting_vott_json method
+ Update CLI to call for tags first then call images then download from blob storage (this way all the images match the tags)

I signed off on the PR

**Review by andrebriggs — APPROVED** [review]

**Comment by samiyaakhtar** [comment]
@andrebriggs I agree with the task breakdown, I can address them in another PR to prevent this one from being bloated
